Markdown reader: support GitHub wiki's internal links (#2923)

blackheaven opened the pull request:
closes https://github.com/jgm/pandoc/issues/2923

Changes overview:

 * Add a `Ext_markdown_github_wikilink` extension
   * Added in the `githubMarkdownExtensions` list
   * Added in the `getAllExtensions` list
 * Add the parser `githubWikiLink` in `Text.Pandoc.Readers.Markdown`
   * Added it high in the `inline` parsers list
 * Add two tests, one for each form

Notes:

 * It has been done as part as ZuriHac 2020
 * It is my first contribution
 * I am available to pair on Discord if needed

Thanks by advance for your feedback.

mb21 commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1735,6 +1736,17 @@ source = do
 linkTitle :: PandocMonad m => MarkdownParser m Text
 linkTitle = quotedTitle '"' <|> quotedTitle '\''
 
+-- Github wiki style link, with optional title
```
maybe add a comment like:

-- syntax documented under https://help.github.com/en/github/building-a-strong-community/editing-wiki-content

mb21 commented on test/Tests/Readers/Markdown.hs:
```diff
@@ -307,6 +307,14 @@ tests = [ testGroup "inline code"
+            para (link "https://example.org" "" (text "https://example.org"))
+          , test markdownGH "link with title" $
+            "[[title|https://example.org]]" =?>
+            para (link "title" "" (text "https://example.org"))
```
maybe add a test for e.g.

    [[Link Text|nameofwikipage]]

and one for:

    [[Name of Page]]

(assuming that's the syntax for pages with spaces in them)

jgm replied:
Another good test (as noted above) wolud be `[[Name of page] and]]`

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1735,6 +1736,17 @@ source = do
+  guardEnabled Ext_markdown_github_wikilink
+  string "[["
+  skipSpaces
+  let allUntil = manyTillChar (satisfy $ const True)
```
`satisfy $ const True` is just `anyChar`, right?

jgm replied:
Also: do we really want anyChar here?  Suppose you have `[[` followed by sixteen paragraphs of text, each separated by a blank line.

jgm replied:
In pandoc's links, the link description can contain inline formatting.  It would be natural to extend that feature here, though I don't know how GitHub's parsing actually works.

mb21 replied:
I just tried in https://github.com/jgm/pandoc/wiki/Sandbox/_edit so yeah, bold and italics appear supported... but not links of course.

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1735,6 +1736,17 @@ source = do
+  skipSpaces
+  let allUntil = manyTillChar (satisfy $ const True)
+  part1 <- allUntil (oneOf "|]")
+  part2 <- (char ']' >> return part1) <|> allUntil (string "]]")
```
`allUntil (string "]]")`: note that you generally want something like `(try (string "]]"))` in this context, because otherwise your parser will fail when it encounters a single `]` with something else after it.  (Maybe that's what you want in this case, though?)

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1735,6 +1736,17 @@ source = do
+  let allUntil = manyTillChar (satisfy $ const True)
+  part1 <- allUntil (oneOf "|]")
+  part2 <- (char ']' >> return part1) <|> allUntil (string "]]")
+  returnF $ B.link part1 "" (B.text $ escapeURI $ trimr part2)
```
In the MediaWiki reader, we insert a special title "wikilink" (if I recall) for wikilinks. This allows them to be distinguished in the AST, so people can do things like transform their URLs (adding a prefix for example).

jgm commented on src/Text/Pandoc/Extensions.hs:
```diff
@@ -118,6 +118,7 @@ data Extension =
     | Ext_literate_haskell    -- ^ Enable literate Haskell conventions
     | Ext_markdown_attribute      -- ^ Interpret text inside HTML as markdown iff
                                   --   container has attribute 'markdown'
+    | Ext_markdown_github_wikilink -- ^ Interpret a markdown wiki link
```
I think I'd favor the name `ext_wikilink`.  This is a pretty generic wikilink format, certainly not GitHub specific.

mb21 replied:
apparently it's inspired by mediawiki.... and [apparently](https://help.github.com/en/github/building-a-strong-community/editing-wiki-content) you can also use mediawiki image syntax:

    [[https://github.com/username/repository/blob/master/img/octocat.png|alt=octocat]]

but yeah, in markdown the link and image syntaxes are also closely related, so why not `ext_wikilink` :)

blackheaven commented:
I have tried to take care of comments:

 * Change the extension name
 * Added documentation
 * Imported the MediaWiki's way to read wikilinks (drop 'Category' part and limit support to italics and strongs)
 * Added tests

Sadly, my tests are failing and I do not see what I am doing wrong, if someone can have a look and see something obvious, I would be grateful.

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1737,15 +1737,15 @@ linkTitle :: PandocMonad m => MarkdownParser m Text
-  part2 <- (char ']' >> return part1) <|> allUntil (string "]]")
-  returnF $ B.link part1 "" (B.text $ escapeURI $ trimr part2)
+  guardEnabled Ext_wikilink
+  try $ string "[["
```
You don't need this `try` or the one on l 1747 because you've got a `try` over the whole `githubWikiLink` parser (which maybe should be named `wikiLinkParser`?)  Having the unnecessary `try`s just hurts performance.

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1737,15 +1737,15 @@ linkTitle :: PandocMonad m => MarkdownParser m Text
+  guardEnabled Ext_wikilink
+  try $ string "[["
+  pagename <- T.unwords . T.words <$> manyChar (noneOf "|]")
+  label <- option (return $ B.text pagename) $ char '|' *> (mconcat <$> many1 (notFollowedBy (string "]]") *> strongOrEmph))
```
Here I think you *do* need a `try`:

`notFollowedBy (try (string "]]"))`

To test this, use something like `[[a|b]and more]]`

blackheaven replied:
good catch, great, thank you!

blackheaven commented:
Any blocking points?

jgm commented on src/Text/Pandoc/Extensions.hs:
```diff
@@ -118,6 +118,7 @@ data Extension =
     | Ext_literate_haskell    -- ^ Enable literate Haskell conventions
     | Ext_markdown_attribute      -- ^ Interpret text inside HTML as markdown iff
                                   --   container has attribute 'markdown'
+    | Ext_wikilink -- ^ Interpret a markdown wiki link
```
Sorry, but on reflection I'm thinking `Ext_wikilinks` would be more consistent with the extension naming style.

jgm commented on test/Tests/Readers/Markdown.hs:
```diff
@@ -307,6 +307,26 @@ tests = [ testGroup "inline code"
+            "[[title|random string]]" =?>
+            para (link "random string" "wikilink" (text "title"))
+          ]
         , testGroup "Headers"
```
Does GitHub (or the markdown parser they use in wikis, which I think may be kramdown) have a suite of tests for this syntax?  If so, it may be worth finding it and seeing if any relevant tests can be added.

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1735,6 +1736,35 @@ source = do
+  string "[["
+  contents <- lookAhead $ manyTillChar anyChar (try $ string "]]")
+  let title = "wikilink"
+  if T.any (== '|') contents
+  then do
+    st <- getState
+    setState $ st{ stateAllowLinks = False }
+    label <- mconcat <$> manyTill inline (char '|')
+    setState $ st{ stateAllowLinks = True }
+    url <- manyTillChar anyChar $ try $ string "]]"
+    return $ B.link url title <$> label
```
This approach seems potentially unreliable.  Example: 
```
[[hello `|` there]]
```
IN this case `contents` will be `` hello `|` there ``, but when we do `manyTill inline (char '|')` we will keep parsing past the `]]`, because the `|` character occurs *within* an inline (a code span).

jgm replied:
This is tricky, and I'm not sure how it should behave, because we don't know whether to treat the content as plain text or inlines until we know whether there's a `|` character, but that `|` character might be part of regular inline content.  I'm really not sure what to do here.

blackheaven replied:
Good point, I have actually borrowed it from the VimWiki parser, we may have the same issue.

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1442,6 +1442,7 @@ inlines1 = mconcat <$> many1 inline
 
 inline :: PandocMonad m => MarkdownParser m (F Inlines)
 inline = choice [ whitespace
+                , githubWikiLink
```
Is there a reason this is here and not further down in the list?
Probably best to put it right before the regular link and image parsers; that should improve performance a bit.

jgm commented:
Worth figuring out whether the GitHub implementation of this allows inline formatting in the description part.  If it's treated as a plain string, maybe we could do the same, and it would solve the problem noted above.  Finding their test suite for this would be nice!

blackheaven commented:
If there is a test suite, it is well hidden.

I have looked for hours, tried babelmark2, no other implementation supports it, no documentation... I begin to doubt of it.

I have tried to make it recursive but it causes issues with autolinks.

blackheaven commented:
Great news! wikilinks are actually way simpler than we expected: https://github.com/blackheaven/pandoc/wiki

blackheaven commented:
it should be good now

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1732,6 +1733,42 @@ source = do
+githubWikiLink = try $ do
+  guardEnabled Ext_wikilinks
+  string "[["
+  inlinedContents <- option (return $ B.fromList [])
```
Note that `B.fromList []` is just `mempty` which would be clearer and faster.

jgm replied:
If you want to get both the inline contents and the raw contents, you can use the `withRaw` combinator (T.P.Parsing).

blackheaven replied:
Actually, I do not know if I did correctly, but it broke the tests

jgm commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1732,6 +1733,42 @@ source = do
+  let allowedInline x = case x of
+                          B.Str _   -> True
+                          B.Space   -> True
+                          B.Link {} -> True
```
Are all markdown links allowed?  Or just bare URIs?
If just bare URIs, then maybe a better approach would be to use a parser that just targets what we need.

jgm replied:
GitHub's documentation says: `[[Link Text|nameofwikipage]]`

So there are two questions.  1) What can go in "Link Text"?  Your experiments appear to reveal that only plain text is allowed, no special formatting.  2) What can go in "nameofwikipage"?  Your experiments show  that a fully qualified URL is allowed, but clearly that's not all -- names of wiki pages are also allowed.  I imagine that just about any plain text could be the name of a wiki page, so presumably we should allow e.g. `[[Link Text|This is my page]]`.  For this we should emit a link with URL `This is my page` but also a special attribute indicating that it's a wikilink, so that users can do transformations on the path.

jgm replied:
(OK, looks like that's what you're doing, from the tests!)

hftf replied:
This is not pertinent to GitHub-flavored Markdown, but it may be informative to note that wikilinks have been growing in popularity due to the explosion of note-taking and (personal) knowledge management tools, which are often based on Markdown or Markdown-ish syntaxes. Some (nonstandard) implementations allow link text to contain inline formatting, and some even allow page titles (and thus link targets) to contain both formatting and nested links.

jgm replied:
Anyway, seems to me we should try to simplify the parser.  Instead of parsing inlines and raw, and then trying to piece things together, why not something like

``` haskell
wikilink = try $ do
  string "[["
  firstPart <- wikiText
  (char '|' *> complexWikilink firstPart) <|> (string "]]" *> return (wikilink with firstPart as link text and (stringify firstPart) as url)

complexWikilink firstPart = do
  url <- stringify <$> wikiText
  string "]]"
  return $ wikilink with firstPart as link text and url as URL

wikiText = many ( whitespace <|> bareURL <|> str <|> endline <|> escapedChar )
```

blackheaven replied:
I have tried you suggestion, but it seems that I miss the point

jgm replied:
It's okay, let me merge your commit, and then I'll see if there's a way to simplify it.

jgm commented:
Sorry I left this sit for so long!  I've made some comments in the code.

blackheaven commented:
> Sorry I left this sit for so long! I've made some comments in the code.

No worries, my sparse time has drastically decreased in the mean time.

jgm commented:
Thanks for this contribution!

jgm commented:
Sorry!  After merging this I realized something that has caused me to revert the commit.

Since this is designed to be used with `gfm`, we should implement the parser as a commonmark extension (commonmark-extensions), rather than here.  The code in pandoc's markdown reader is NOT used when parsing `gfm`. I presume that most users who will want this feature will want to use it with `gfm`, so I think we should wait.

I'm sorry I didn't think of this earlier, before you put the time into this PR.

Writing a commonmark-hs extension is an entirely different thing, and I need to think more about how it would be done with this syntax.

jgm commented:
The work you've done in constructing the tests and in the extensions module can be re-used once we have the new commonmark-hs module in place.

jgm commented:
I've started to add a `wikilinks` extension in commonmark-hs.

blackheaven commented:
Sure, no problem

marcelotto commented:
This is a really exciting extension. I was already about to submit a feature request when I came across this PR. I just want to undermine that this extension isn't only useful for Github Wiki pages. Using Wikilinks in Markdown is getting fairly popular recently and presumably even more in the future with the recent raise of local-first, Markdown-based knowledge management tools like [Obsidian](https://obsidian.md/), [Logseq](https://github.com/logseq/logseq), [Dendron](https://github.com/dendronhq/dendron) to name just a few.

jgm commented:
I've run into a problem implementing this.
There are two incompatible syntaxes (different orders) out there: see
https://github.com/jgm/commonmark-hs/issues/69
Not quite sure what to do about it.

hftf commented:
There are now many incompatible implementations and interpretations of "wikilinks" in existence, which is partly due to the recent growth of note-taking tools that [I](https://github.com/jgm/pandoc/pull/6458#discussion_r552125324) and others have mentioned above.

I foresee that many users of pandoc are going to want to operate on documents containing wikilinks of all kinds. This is a challenge.

Perhaps an encompassing solution would be "DIY." Instead of favoring one implementation out of many for a single extension called `wikilinks`, Pandoc could implement various small pieces of functionality commonly found in wikilink implementations, such as a pair of mutually exclusive options to handle either `[[target|link text]]` or `[[link text|target]]` (`wikilink-piped-link-text-on-right` and `-left`), or to allow or disallow nested inline formatting. Particularly well-known implementations could be aliased to combinations of options (e.g. `wikilinks-gfm` is an alias of `wikilinks-option1+wikilinks-option3` etc.).

marcelotto commented:
Sorry @hftf, didn't notice your similar remarks. I guess you're right, it's a challenging problem which ideally would be solved in the most flexible way you've outlined, but I can image that this can lead to a maintainability nightmare. From what I can see two alternative extensions should be sufficient if the extensions are limited on capturing the pure links: One for classical Wikilinks (`[[link]]` and `[[link|label]]`) and one for Github-flavoured Wikilinks (`[[link]]` and `[[label|link]]`). Regarding the additional system-specific features on top of that (inline formatting etc.), either people with portability in mind are hesitant on their usage in the first place or have  with these extensions at least the option to capture the links with a filter and do the additional interpretation on their own.

One remark regarding the understandably frustrated comment

> This is how it's done in GitHub wikis. I don't know why everyone has to reinvent the wheel in incompatible ways!

by @jgm in the [commonmark-hs issue](https://github.com/jgm/commonmark-hs/issues/69). It should be noted that GitHub were the ones breaking with the established way. The other tools are using the classical Wikilink-style introduced by UseModWiki and used in almost all Wiki systems [1] and in all the new systems from the new generation of knowledge management tools (as far as I can see).

jgm commented:
Fair point: GitHub was never great about looking at prior art in making markdown extensions.
